### PR TITLE
Share build cache configuration instead of build cache service

### LINKED ISSUE #4216: NullPointerException in DefaultPersistentDirectoryStore when reading/writing from build cache
We're seeing a number of NullPointerExceptions while populating the build cache.

We're using composite builds, so I can't attach a build scan.

This is using Gradle 4.5, with both org.gradle.parallel and org.gradle.configureondemand set to true. 

I don't believe the errors are impacting the results of the build, however they are likely hurting the effectiveness of the cache. 

```text
Failed to store cache entry df0ce2b67434b3baa11047ea6dbbe708 for task ':sdk:nz.co.tvnz.frame.commons:test'
java.lang.NullPointerException
        at org.gradle.cache.internal.DefaultPersistentDirectoryStore.withFileLock(DefaultPersistentDirectoryStore.java:163)
        at org.gradle.cache.internal.DefaultCacheFactory$ReferenceTrackingCache.withFileLock(DefaultCacheFactory.java:192)
        at org.gradle.caching.local.internal.DirectoryBuildCacheService.withTempFile(DirectoryBuildCacheService.java:176)
        at org.gradle.caching.internal.controller.DefaultBuildCacheController.store(DefaultBuildCacheController.java:208)
        at org.gradle.caching.internal.controller.RootBuildCacheControllerRef$CloseShieldBuildCacheController.store(RootBuildCacheControllerRef.java:59)
        at org.gradle.api.internal.tasks.execution.SkipCachedTaskExecuter.execute(SkipCachedTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.execute(ResolveBuildCacheKeyExecuter.java:63)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:623)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:578)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by wolfs**
Sharing the service causes problems when they are shut down - sometimes
the root service is shut down even if one of the included builds is
still running and using it.

This provides a fix for #4216.

### CODE REVIEW

File: subprojects/build-cache/src/integTest/groovy/org/gradle/caching/configuration/BuildCacheCompositeConfigurationIntegrationTest.groovy [hunk 1]
```diff
@@ -106,10 +109,55 @@ class BuildCacheCompositeConfigurationIntegrationTest extends AbstractIntegratio
...
+            println gradle
+            gradle.startParameter.setTaskNames(['build'])
+            processResources {
+                dependsOn gradle.includedBuild('i1').task(':processResources')
+                dependsOn gradle.includedBuild('i2').task(':processResources')
+            }
+        """
+
+        expect:
+        succeeds '--parallel'
```
Comment (adammurdoch) [inline subprojects/build-cache/src/integTest/groovy/org/gradle/caching/configuration/BuildCacheCompositeConfigurationIntegrationTest.groovy]: you don't need `--parallel` as the tasks for included builds are run in parallel anyway.

**Comment by ldaley** [comment]
This seems like the wrong solution to me. I think we should fix the lifecycling instead.

For example, this would complicate a solution for pre-emptively opening connections to the remote cache to amortize the connection cost.

File: subprojects/build-cache/src/integTest/groovy/org/gradle/caching/configuration/BuildCacheCompositeConfigurationIntegrationTest.groovy [hunk 2]
```diff
@@ -96,7 +99,7 @@ class BuildCacheCompositeConfigurationIntegrationTest extends AbstractIntegratio
 
         and:
         def finalizeOps = operations.all(FinalizeBuildCacheConfigurationBuildOperationType)
-        finalizeOps.size() == 2
+        finalizeOps.size() == 5
```
Comment (ldaley) [inline subprojects/build-cache/src/integTest/groovy/org/gradle/caching/configuration/BuildCacheCompositeConfigurationIntegrationTest.groovy]: This is a break for build scans. 

We either need to roll this back (e.g. don't fire this event for non root builds), or we have to update build scans to ignore these events.

**Comment by ldaley** [comment]
I've just realised that `buildFinished` for the root build firing before all other builds have finished is very detrimental to build scans. 

I think we should just not fire the root buildFinished until the other builds have completed, as per #4199.

**Comment by wolfs** [comment]
Closed in favor of #4249

File: subprojects/build-cache/src/integTest/groovy/org/gradle/caching/configuration/BuildCacheCompositeConfigurationIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/build-cache/src/integTest/groovy/org/gradle/caching/configuration/BuildCacheCompositeConfigurationIntegrationTest.groovy]: Good catch, @adammurdoch. Removed.

**Pull request closed without merging**